gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/UserDefinedFunction.java

```diff
@@ -0,0 +1,158 @@
+   */
+  public enum Type {
+    INLINE,
+    FROM_URI
+  }
+
+  private final Type type;
+  private final String functionDefinition;
+
+  UserDefinedFunction(Type type, String functionDefinition) {
+    this.type = type;
+    this.functionDefinition = functionDefinition;
+  }
+
+  public Type type() {
+    return type;
+  }
+
+  /**
+   * Returns function's definition. If {@link #type()} is {@link Type#INLINE} this method returns
+   * a code blob. If {@link #type()} is {@link Type#FROM_URI} this method returns a Google Cloud
+   * Storage URI (e.g. gs://bucket/path).
+   */
+  public String functionDefinition() {
+    return functionDefinition;
+  }
+
+  /**
+   * A Google Cloud BigQuery user-defined function, as a code blob.
+   */
+  public static final class InlineFunction extends UserDefinedFunction {
+
+    private static final long serialVersionUID = 1083672109192091686L;
+
+    /**
+     * Creates a Google Cloud BigQuery user-defined function given a code blob.
+     */
+    public InlineFunction(String inlineCode) {
+      super(Type.INLINE, inlineCode);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("inlineCode", functionDefinition()).toString();
+    }
+
+    @Override
+    public com.google.api.services.bigquery.model.UserDefinedFunctionResource toPb() {
+      return new com.google.api.services.bigquery.model.UserDefinedFunctionResource()
+          .setInlineCode(functionDefinition());
+    }
+  }
+
+  /**
+   * A Google Cloud BigQuery user-defined function, as an URI to Google Cloud Storage.
+   */
+  public static final class UriFunction extends UserDefinedFunction {
+
+    private static final long serialVersionUID = 4660331691852223839L;
+
+    /**
+     * Creates a Google Cloud BigQuery user-defined function given a Google Cloud Storage URI (e.g.
+     * gs://bucket/path).
+     */
+    public UriFunction(String functionUri) {
+      super(Type.FROM_URI, functionUri);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("functionUri", functionDefinition()).toString();
+    }
+
+    @Override
+    public com.google.api.services.bigquery.model.UserDefinedFunctionResource toPb() {
+      return new com.google.api.services.bigquery.model.UserDefinedFunctionResource()
+          .setResourceUri(functionDefinition());
+    }
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(type, functionDefinition);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return obj instanceof UserDefinedFunction
+        && Objects.equals(toPb(), ((UserDefinedFunction) obj).toPb());
+  }
+
+  public abstract com.google.api.services.bigquery.model.UserDefinedFunctionResource toPb();
+
+  /**
+   * Creates a Google Cloud BigQuery user-defined function given a code blob.
+   */
+  public static UserDefinedFunction inline(String functionDefinition) {
+    return new InlineFunction(functionDefinition);
```

Comment:
Also, in its current state I am not sure there is even a value in exposing the specific sub-classes 'UriFunction` and `UserDefinedFunction`.